### PR TITLE
RavenDB-18727 - Sharding - Revisions - Handlers

### PR DESCRIPTION
**Pull request by shiranshalom**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18727/Sharding-Revisions-Handlers
https://issues.hibernatingrhinos.com/issue/RavenDB-18645/Sharding-Studio-Move-handlers-to-processors-implement-missing
https://issues.hibernatingrhinos.com/issue/RavenDB-19066/Sharding-handle-the-Etag-to-return-not-modified-status

### Additional description

Handle  `NotModified` for:
- `ShardedRevisionsHandlerProcessorForGetRevisions`
- `GetFieldsAsync`
- `ShardedStudioCollectionsHandlerProcessorForPreviewCollection`

### Type of change

- Task

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs [hunk 1]
```diff
@@ -72,31 +74,33 @@ protected override async ValueTask GetRevisionsAsync(TransactionOperationContext
                 cmd = new GetRevisionsCommand(id, start, pageSize, metadataOnly);
             }
 
-            string etag = null; //TODO
-            if (NotModified(etag))
+            int shardNumber = RequestHandler.DatabaseContext.GetShardNumber(context, id);
+            var result = await RequestHandler.ExecuteSingleShardAsync(context, cmd, shardNumber, token);
+            var array = result.Results.Items.Select(x => ((BlittableJsonReaderObject)x).Clone(context)).ToArray();
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs]: do you know why are we cloning here?

**Review by ppekrol — COMMENTED** [review]
I do not see a test?

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs (hunk 1 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs]: @lastav5 can you answer this? :-)

**Comment by shiranshalom** [comment]
> I do not see a test?

Done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs (hunk 1 above)
Comment (lastav5) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs]: Not sure.. Probably a mistake. Doesn't seem like the cloning is needed now, if it ever was

Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs]: Done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs [hunk 2]
```diff
@@ -36,15 +37,16 @@ protected override async ValueTask GetRevisionByChangeVectorAsync(TransactionOpe
                 return;
             }
 
-            string etag = null; //TODO
-            if (NotModified(etag))
+            var revisions = ShardResultConverter.BlittableToDocumentsListConverter(res);
+            var actualEtag = ComputeHttpEtags.ComputeEtagForRevisions(revisions);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs]: I wonder if we need to work so hard... would xor on the etag headers from the requests would be easier?

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs [hunk 3]
```diff
@@ -72,31 +74,33 @@ protected override async ValueTask GetRevisionsAsync(TransactionOperationContext
...
+            var array = result.Results.Items.Select(x => (BlittableJsonReaderObject)x).ToArray();
+
+            string actualChangeVector;
+            if (array.Length == 0)
+                actualChangeVector = "";
+            else
+            {
+                var revision = ShardResultConverter.BlittableToDocumentConverter(array[0]);
+                actualChangeVector = revision.ChangeVector;
+            }
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs]: why can't we just use the existing the header from the command?

File: src/Raven.Server/Documents/Sharding/Streaming/ShardResultConverter.cs [hunk 4]
```diff
@@ -8,14 +8,30 @@ public static class ShardResultConverter
     {
         public static Document BlittableToDocumentConverter(BlittableJsonReaderObject json)
         {
+            if (json == null)
+                return null;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Streaming/ShardResultConverter.cs]: when can this happened?

Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Streaming/ShardResultConverter.cs]: Removed

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs (hunk 2 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs]: Not sure I follow.. We don't have the headers here...

Comment (karmeli87) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs]: You can get it (or maybe compute it?) from the `ShardedGetRevisionsByChangeVectorsOperation` above... no?

**Comment by shiranshalom** [comment]
PR message has been updated.

File: src/Raven.Server/Documents/Sharding/Executors/AbstractExecutor.cs [hunk 5]
```diff
@@ -77,6 +80,9 @@ public Task<TResult> ExecuteParallelForAllAsync<TResult>(IShardedOperation<TResu
             for (var index = 0; index < position; index++)
             {
                 var command = commands[index];
+                if (etagFromCommandsResponse != null)
+                    etagFromCommandsResponse[index] = commands[index].Command.Etag;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Executors/AbstractExecutor.cs]: please check `IShardedReadOperation`

**Review by karmeli87 — CHANGES_REQUESTED** [review]
I believe that most of the changes here should be handled via the `IShardedReadOperation`

File: src/Raven.Server/Documents/Sharding/Executors/AbstractExecutor.cs (hunk 5 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Executors/AbstractExecutor.cs]: Done

File: test/FastTests/Server/Documents/Revisions/RevisionsTests.cs [hunk 6]
```diff
@@ -2146,6 +2147,120 @@ public async Task CanGetRevisionsCountForAsync(Options options)
...
+                    var etagFromResponse = response.Headers.ETag?.Tag;
+
+                    request = revisionsCommand.CreateRequest(ctx, serverNode, out url);
+                    request.RequestUri = new UriBuilder(url).Uri;
+                    request.Headers.IfNoneMatch.ParseAdd(etagFromResponse);
+
+                    response = await re.HttpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, CancellationToken.None);
+                    
+                    Assert.True(response.StatusCode == HttpStatusCode.NotModified);
+                }
```
Comment (karmeli87) [inline test/FastTests/Server/Documents/Revisions/RevisionsTests.cs]: Would it be simpler to use the `GetRevisionOperation` and hook to the `OnSucceedRequest` ?

File: test/FastTests/Server/Documents/Revisions/RevisionsTests.cs [hunk 7]
```diff
@@ -2146,6 +2147,120 @@ public async Task CanGetRevisionsCountForAsync(Options options)
...
+
+                        var etagFromResponse = response.Headers.ETag?.Tag;
+
+                        request = revisionsCommand.CreateRequest(ctx, serverNode, out url);
+                        request.RequestUri = new UriBuilder(url).Uri;
+                        request.Headers.IfNoneMatch.ParseAdd(etagFromResponse);
+
+                        response = await re.HttpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, CancellationToken.None);
+
+                        Assert.True(response.StatusCode == HttpStatusCode.NotModified);
```
Comment (karmeli87) [inline test/FastTests/Server/Documents/Revisions/RevisionsTests.cs]: same

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs [hunk 8]
```diff
@@ -72,31 +72,25 @@ protected override async ValueTask GetRevisionsAsync(TransactionOperationContext
                 cmd = new GetRevisionsCommand(id, start, pageSize, metadataOnly);
             }
 
-            string etag = null; //TODO
-            if (NotModified(etag))
+            int shardNumber = RequestHandler.DatabaseContext.GetShardNumber(context, id);
+            var result = await RequestHandler.ExecuteSingleShardAsync(context, cmd, shardNumber, token);
+
+            string actualChangeVector = cmd.Etag;
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs]: let's rename to `actualEtag`

File: test/FastTests/Server/Documents/Revisions/RevisionsTests.cs (hunk 6 above)
Comment (shiranshalom) [reply test/FastTests/Server/Documents/Revisions/RevisionsTests.cs]: Done

File: test/FastTests/Server/Documents/Revisions/RevisionsTests.cs (hunk 7 above)
Comment (shiranshalom) [reply test/FastTests/Server/Documents/Revisions/RevisionsTests.cs]: Done

File: src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs (hunk 8 above)
Comment (shiranshalom) [reply src/Raven.Server/Documents/Sharding/Handlers/Processors/Revisions/ShardedRevisionsHandlerProcessorForGetRevisions.cs]: Done

File: test/SlowTests/Client/Operations/StudioOperationsTests.cs [hunk 9]
```diff
@@ -181,6 +184,102 @@ public async Task CanGetStudioCollectionFields(Options options)
...
+                    var etagFromResponse = response.Headers.ETag?.Tag;
+
+                    command = new GetCollectionFieldsCommand("User", "");
+                    request = command.CreateRequest(ctx, serverNode, out url);
+                    request.RequestUri = new UriBuilder(url).Uri;
+                    request.Headers.IfNoneMatch.ParseAdd(etagFromResponse);
+
+                    response = await re.HttpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, CancellationToken.None);
+
+                    Assert.True(response.StatusCode == HttpStatusCode.NotModified);
```
Comment (karmeli87) [inline test/SlowTests/Client/Operations/StudioOperationsTests.cs]: same?

File: test/SlowTests/Client/Operations/StudioOperationsTests.cs [hunk 10]
```diff
@@ -181,6 +184,102 @@ public async Task CanGetStudioCollectionFields(Options options)
...
+                    var etagFromResponse = response.Headers.ETag?.Tag;
+
+                    request = new HttpRequestMessage { Method = HttpMethod.Get };
+                    url = $"{serverNode.Url}/databases/{serverNode.Database}/studio/collections/preview";
+                    request.RequestUri = new UriBuilder(url).Uri;
+                    request.Headers.IfNoneMatch.ParseAdd(etagFromResponse);
+
+                    response = await re.HttpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, CancellationToken.None);
+
+                    Assert.True(response.StatusCode == HttpStatusCode.NotModified);
```
Comment (karmeli87) [inline test/SlowTests/Client/Operations/StudioOperationsTests.cs]: same?

File: test/SlowTests/Client/Operations/StudioOperationsTests.cs (hunk 9 above)
Comment (shiranshalom) [reply test/SlowTests/Client/Operations/StudioOperationsTests.cs]: We don't have an operation for that...

Comment (karmeli87) [reply test/SlowTests/Client/Operations/StudioOperationsTests.cs]: `GetCollectionFieldsOperation`? no? we always can create one :)

Comment (shiranshalom) [reply test/SlowTests/Client/Operations/StudioOperationsTests.cs]: Done

File: test/SlowTests/Client/Operations/StudioOperationsTests.cs (hunk 10 above)
Comment (shiranshalom) [reply test/SlowTests/Client/Operations/StudioOperationsTests.cs]: Done

File: src/Raven.Client/Documents/Session/DocumentSessionCollectionOperationsAsync.cs [hunk 11]
```diff
@@ -0,0 +1,78 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using Raven.Client.Documents.Session.Operations;
+using Raven.Client.Http;
+using Sparrow.Json;
+
+namespace Raven.Client.Documents.Session
+{
+    public class DocumentSessionCollectionOperations : AdvancedSessionExtensionBase, ICollectionSessionOperations
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Session/DocumentSessionCollectionOperationsAsync.cs]: I don't think we need all this

File: src/Raven.Client/Documents/Commands/PreviewCollectionCommand.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Net.Http;
+using System.Text;
+using Raven.Client.Http;
+using Sparrow.Json;
+
+namespace Raven.Client.Documents.Commands
+{
+    public class PreviewCollectionCommand : RavenCommand<BlittableJsonReaderObject>
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Commands/PreviewCollectionCommand.cs]: internal

File: src/Raven.Client/Documents/Commands/GetCollectionFieldsCommand.cs [hunk 13]
```diff
@@ -0,0 +1,62 @@
+﻿using System;
+using System.Net.Http;
+using System.Text;
+using Raven.Client.Http;
+using Sparrow.Json;
+
+namespace Raven.Client.Documents.Commands
+{
+    public class GetCollectionFieldsCommand : RavenCommand<BlittableJsonReaderObject>
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Commands/GetCollectionFieldsCommand.cs]: internal

File: src/Raven.Client/Documents/Session/Operations/PreviewCollectionOperation.cs [hunk 14]
```diff
@@ -0,0 +1,27 @@
+﻿using Raven.Client.Documents.Commands;
+using Sparrow.Json;
+
+namespace Raven.Client.Documents.Session.Operations
+{
+    public class PreviewCollectionOperation
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Session/Operations/PreviewCollectionOperation.cs]: internal

File: src/Raven.Client/Raven.Client.csproj [hunk 15]
```diff
@@ -137,6 +137,7 @@
     <PackageReference Include="Microsoft.AspNetCore.JsonPatch" Version="7.0.0" />
     <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="7.0.0" />
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
+    <PackageReference Include="Microsoft.Extensions.Primitives" Version="2.1.1" />
```
Comment (karmeli87) [inline src/Raven.Client/Raven.Client.csproj]: what is that?

**Review by karmeli87 — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Commands/PreviewCollectionCommand.cs (hunk 12 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Commands/PreviewCollectionCommand.cs]: Done

File: src/Raven.Client/Documents/Commands/GetCollectionFieldsCommand.cs (hunk 13 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Commands/GetCollectionFieldsCommand.cs]: Done

File: src/Raven.Client/Documents/Session/Operations/PreviewCollectionOperation.cs (hunk 14 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Session/Operations/PreviewCollectionOperation.cs]: Done

File: src/Raven.Client/Raven.Client.csproj (hunk 15 above)
Comment (shiranshalom) [reply src/Raven.Client/Raven.Client.csproj]: Done

File: src/Raven.Client/Documents/Session/DocumentSessionCollectionOperationsAsync.cs (hunk 11 above)
Comment (shiranshalom) [reply src/Raven.Client/Documents/Session/DocumentSessionCollectionOperationsAsync.cs]: Done

File: test/FastTests/Server/Documents/Revisions/RevisionsTests.cs [hunk 16]
```diff
@@ -2146,6 +2137,93 @@ public async Task CanGetRevisionsCountForAsync(Options options)
...
+                {
+                    var company2 = await session.LoadAsync<Company>(company.Id);
+                    company2.Name = "Hibernating Rhinos";
+                    await session.SaveChangesAsync();
+                }
+
+                using (var session = store.OpenSession())
+                {
+                    HttpStatusCode status = default;
+                    session.Advanced.RequestExecutor.OnSucceedRequest += (_, args) => { status = args.Response.StatusCode; };
```
Comment (karmeli87) [inline test/FastTests/Server/Documents/Revisions/RevisionsTests.cs:2164]: we need to check the url here

File: test/FastTests/Server/Documents/Revisions/RevisionsTests.cs [hunk 17]
```diff
@@ -2146,6 +2137,93 @@ public async Task CanGetRevisionsCountForAsync(Options options)
...
+                using (var session = store.OpenAsyncSession())
+                {
+                    var revisionsMetadata = await session.Advanced.Revisions.GetMetadataForAsync(id);
+                    Assert.Equal(11, revisionsMetadata.Count);
+
+                    var changeVectors = revisionsMetadata.Select(x => x.GetString(Constants.Documents.Metadata.ChangeVector)).ToList();
+                    changeVectors.Add("NotExistsChangeVector");
+
+                    HttpStatusCode status = default;
+                    session.Advanced.RequestExecutor.OnSucceedRequest += (_, args) => { status = args.Response.StatusCode; };
```
Comment (karmeli87) [inline test/FastTests/Server/Documents/Revisions/RevisionsTests.cs:2212]: and here

File: test/SlowTests/Client/Operations/StudioOperationsTests.cs [hunk 18]
```diff
@@ -181,6 +183,74 @@ public async Task CanGetStudioCollectionFields(Options options)
...
+
+                    var o3 = new NoFieldsClass();
+                    session.Store(o3, "user/E-3");
+                    session.Advanced.GetMetadataFor(o3)[Constants.Documents.Metadata.Collection] = "User";
+                    session.SaveChanges();
+                }
+                using (var session = store.OpenAsyncSession())
+                {
+                    HttpStatusCode status = default;
+                    session.Advanced.RequestExecutor.OnSucceedRequest += (_, args) => { status = args.Response.StatusCode; };
```
Comment (karmeli87) [inline test/SlowTests/Client/Operations/StudioOperationsTests.cs:210]: and here

File: test/SlowTests/Client/Operations/StudioOperationsTests.cs [hunk 19]
```diff
@@ -181,6 +183,74 @@ public async Task CanGetStudioCollectionFields(Options options)
...
+                using (var session = store.OpenSession())
+                {
+                    session.Store(new User {Name = "SomeDoc"});
+                    session.SaveChanges();
+                }
+
+                using (var session = store.OpenAsyncSession())
+                {
+                    HttpStatusCode status = default;
+                    session.Advanced.RequestExecutor.OnSucceedRequest += (_, args) => { status = args.Response.StatusCode; };
```
Comment (karmeli87) [inline test/SlowTests/Client/Operations/StudioOperationsTests.cs:239]: and here

File: test/FastTests/Server/Documents/Revisions/RevisionsTests.cs (hunk 16 above)
Comment (shiranshalom) [reply test/FastTests/Server/Documents/Revisions/RevisionsTests.cs:2164]: done

File: test/FastTests/Server/Documents/Revisions/RevisionsTests.cs (hunk 17 above)
Comment (shiranshalom) [reply test/FastTests/Server/Documents/Revisions/RevisionsTests.cs:2212]: done

File: test/SlowTests/Client/Operations/StudioOperationsTests.cs (hunk 18 above)
Comment (shiranshalom) [reply test/SlowTests/Client/Operations/StudioOperationsTests.cs:210]: done

File: test/SlowTests/Client/Operations/StudioOperationsTests.cs (hunk 19 above)
Comment (shiranshalom) [reply test/SlowTests/Client/Operations/StudioOperationsTests.cs:239]: done

**Pull request merged**